src/platform/backends/shared/base_virtual_machine.h

```diff
@@ -29,26 +30,114 @@
 #include <QRegularExpression>
 #include <QString>
 
-namespace mp = multipass;
-namespace mpl = multipass::logging;
-namespace mpu = multipass::utils;
+#include <memory>
+#include <mutex>
+#include <unordered_map>
 
 namespace multipass
 {
 class BaseVirtualMachine : public VirtualMachine
 {
 public:
-    BaseVirtualMachine(VirtualMachine::State state, const std::string& vm_name) : VirtualMachine(state, vm_name){};
-    BaseVirtualMachine(const std::string& vm_name) : VirtualMachine(vm_name){};
+    BaseVirtualMachine(VirtualMachine::State state, const std::string& vm_name, const Path& instance_dir);
+    BaseVirtualMachine(const std::string& vm_name, const Path& instance_dir);
 
     std::vector<std::string> get_all_ipv4(const SSHKeyProvider& key_provider) override;
     std::unique_ptr<MountHandler> make_native_mount_handler(const SSHKeyProvider* ssh_key_provider,
                                                             const std::string& target,
-                                                            const multipass::VMMount& mount) override
+                                                            const VMMount& mount) override
     {
         throw NotImplementedOnThisBackendException("native mounts");
     };
+
+    SnapshotVista view_snapshots() const override;
+    int get_num_snapshots() const noexcept override;
+
+    std::shared_ptr<const Snapshot> get_snapshot(const std::string& name) const override;
+    std::shared_ptr<const Snapshot> get_snapshot(int index) const override;
+    std::shared_ptr<Snapshot> get_snapshot(const std::string& name) override;
+    std::shared_ptr<Snapshot> get_snapshot(int index) override;
+
+    // TODO: the VM should know its directory, but that is true of everything in its VMDescription; pulling that from
+    // derived classes is a big refactor
+    std::shared_ptr<const Snapshot> take_snapshot(const VMSpecs& specs,
+                                                  const std::string& snapshot_name,
+                                                  const std::string& comment) override;
+    void rename_snapshot(const std::string& old_name, const std::string& new_name) override;
+    void delete_snapshot(const std::string& name) override;
+    void restore_snapshot(const std::string& name, VMSpecs& specs) override;
+    void load_snapshots() override;
+    std::vector<std::string> get_childrens_names(const Snapshot* parent) const override;
+    int get_snapshot_count() const override;
+
+protected:
+    virtual std::shared_ptr<Snapshot> make_specific_snapshot(const QString& filename) = 0;
+    virtual std::shared_ptr<Snapshot> make_specific_snapshot(const std::string& snapshot_name,
+                                                             const std::string& comment,
+                                                             const VMSpecs& specs,
+                                                             std::shared_ptr<Snapshot> parent) = 0;
```

Comment:
We can. I tend to like this approach better, even if it adds a little more duplication, just because it makes it obvious that this is an interface that is meant to be implemented by descendants and because it is easier to find what is or isn't implemented for a given backend. But I don't have a strong opinion and since the rest is done as you describe, I will make it adhere to that.